src/package.rs

```diff
@@ -603,160 +484,206 @@ impl Package {
                     PackageOutput::Zone { .. } => {
                         // Zone images must explicitly label all destination paths
                         // as within "root/".
-                        archive_path(&dst)?
+                        zone_archive_path(&dst)?
                     }
                     PackageOutput::Tarball => dst,
                 };
 
                 if entry.file_type().is_dir() {
-                    archive.append_dir(&dst, ".")?;
+                    inputs
+                        .0
+                        .push(BuildInput::AddDirectory(TargetDirectory(dst)));
                 } else if entry.file_type().is_file() {
-                    archive
-                        .append_path_with_name_async(entry.path(), &dst)
-                        .await
-                        .context(format!(
-                            "Failed to add file '{}' to '{}'",
-                            entry.path().display(),
-                            dst.display()
-                        ))?;
+                    let src = <&Utf8Path>::try_from(entry.path())?;
+                    inputs.0.push(BuildInput::add_file(MappedPath {
+                        from: src.to_path_buf(),
+                        to: dst,
+                    })?);
                 } else {
                     panic!(
                         "Unsupported file type: {:?} for {:?}",
                         entry.file_type(),
                         entry
                     );
                 }
-                progress.increment(1);
             }
         }
-        Ok(())
+
+        Ok(inputs)
     }
 
-    async fn add_rust<W: std::io::Write + Send>(
+    fn get_all_inputs(
         &self,
-        progress: &impl Progress,
-        archive: &mut Builder<W>,
-    ) -> Result<()> {
+        package_name: &str,
+        target: &Target,
+        output_directory: &Utf8Path,
+        zoned: bool,
+        version: Option<&semver::Version>,
+    ) -> Result<BuildInputs> {
+        let mut all_paths = BuildInputs::new();
+
+        // For all archive formats, the version comes first
+        all_paths
+            .0
+            .push(self.get_version_input(package_name, version));
+
+        match &self.source {
+            PackageSource::Local { paths, .. } => {
+                all_paths.0.extend(self.get_paths_inputs(target, paths)?.0);
+                all_paths.0.extend(self.get_rust_inputs()?.0);
+                all_paths
+                    .0
+                    .extend(self.get_blobs_inputs(output_directory, zoned)?.0);
+            }
+            PackageSource::Composite { packages } => {
+                for component_package in packages {
+                    all_paths.0.push(BuildInput::AddPackage(TargetPackage(
+                        output_directory.join(component_package),
+                    )));
+                }
+            }
+            _ => {
+                return Err(anyhow!(
```

Comment:
another `bail!` possibility

Reply:
Done in c6e02731eee201348d38e29390bd6ef60fd30b81